File: app/src/main/java/com/cere/csplayer/content/MusicManager.java
package com.cere.csplayer.content;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Handler;

import com.cere.csplayer.R;
import com.cere.csplayer.activity.MainActivity;
import com.cere.csplayer.data.SQLite;
import com.cere.csplayer.entity.Music;
import com.cere.csplayer.entity.Play;
import com.cere.csplayer.until.DialogUtils;
import com.hjq.permissions.OnPermission;
import com.hjq.permissions.Permission;
import com.hjq.permissions.XXPermissions;
import com.litesuits.orm.db.assit.WhereBuilder;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by CheRevir on 2019/10/22
 */
public class MusicManager implements MusicScanner.OnScanDoneListener, PlayBuilder.OnBuilderDonListener {
    private Context mContext;
    private Handler mHandler;
    private OnLoadDoneListener mOnLoadDoneListener;
    private ArrayList<Music> mMusics;
    private ArrayList<Play> mPlays;

    public MusicManager(MainActivity activity) {
        mContext = activity;
        mOnLoadDoneListener = activity;
        mHandler = new Handler(mContext.getMainLooper());
    }

    public void load() {
        String[] permissions = new String[]{Permission.READ_EXTERNAL_STORAGE, Permission.WRITE_EXTERNAL_STORAGE, Permission.SYSTEM_ALERT_WINDOW};
        if (XXPermissions.isHasPermission(mContext, permissions)) {
            mMusics = SQLite.getLiteOrm(mContext).query(Music.class);
            if (!mMusics.isEmpty()) {
                mPlays = SQLite.getLiteOrm(mContext).query(Play.class);
                mOnLoadDoneListener.onMusicLoadDone(mMusics);
                if (!mPlays.isEmpty()) {
                    mOnLoadDoneListener.onPlayLoadDone(mPlays);
                } else {
                    reBuildPlays();
                }
            } else {
                reScanMusics();
            }
        } else {
            XXPermissions.with((MainActivity) mContext).constantRequest().permission(permissions).request(new OnPermission() {
                @Override
                public void hasPermission(List<String> granted, boolean isAll) {
                    if (isAll) {
                        load();
                    }
                }

                @Override
                public void noPermission(List<String> denied, boolean quick) {
                    DialogUtils.getAlertDialog(mContext, "提示", "没有权限将无法扫描音乐文件。", R.mipmap.ic_launcher)
                            .setPositiveButton("去设置打开", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    XXPermissions.gotoPermissionSettings(mContext);
                                }
                            }).show();
                }
            });
        }
    }

    @Override
    public void onScanDone(ArrayList<Music> list) {
        if (!list.isEmpty()) {
            this.mMusics = list;
            mOnLoadDoneListener.onMusicLoadDone(list);
            reBuildPlays();
            mHandler.post(new Runnable() {
                @Override
                public void run() {
                    SQLite.getLiteOrm(mContext).delete(Music.class);
                    SQLite.getLiteOrm(mContext).save(mMusics);
                }
            });
        } else {
            DialogUtils.getAlertDialog(mContext, "提示", "没有扫描到音乐文件。", R.mipmap.ic_launcher).setPositiveButton("确定", null).show();
        }
    }

    @Override
    public void onBuildDone(final ArrayList<Play> list) {
        this.mPlays = list;
        mOnLoadDoneListener.onPlayLoadDone(list);
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                SQLite.getLiteOrm(mContext).delete(Play.class);
                SQLite.getLiteOrm(mContext).save(list);
            }
        });
    }

    public void reScanMusics() {
        new MusicScanner(mContext, this).execute();
    }

    public void reBuildPlays() {
        Play play = new Play();
        String data = ((MainActivity) mContext).getData();
        if (mPlays != null && !data.isEmpty()) {
            play.setPosition(getPosition(data));
        } else {
            play.setPosition(0);
        }
        new PlayBuilder(mContext, this, play).execute(mMusics);
    }

    public void deleteMusic(final String path) {
        Music music = new Music();
        music.setPath(path);
        mMusics.remove(music);
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                WhereBuilder builder = new WhereBuilder(Music.class);
                builder.equals("path", path);
                SQLite.getLiteOrm(mContext).delete(builder);
            }
        });
    }

    public void deletePlay(final int position) {
        Play play = new Play();
        play.setPosition(position);
        mPlays.remove(position);
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                WhereBuilder builder = new WhereBuilder(Play.class);
                builder.equals("position", position);
                SQLite.getLiteOrm(mContext).delete(builder);
            }
        });
    }

    public int getPosition(String path) {
        Music music = new Music();
        music.setPath(path);
        return mMusics.indexOf(music);
    }

    public int getPosition(int position) {
        Play play = new Play();
        play.setPosition(position);
        return mPlays.indexOf(play);
    }

    public Music getMusic(String path) {
        return mMusics.get(getPosition(path));
    }

    public Play getPlay(int position) {
        return mPlays.get(position);
    }

    public ArrayList<Music> getMusics() {
        return mMusics;
    }

    public ArrayList<Play> getPlays() {
        return mPlays;
    }

    public interface OnLoadDoneListener {
        void onMusicLoadDone(ArrayList<Music> list);

        void onPlayLoadDone(ArrayList<Play> list);
    }
}

File: app/src/main/java/com/cere/csplayer/content/MusicManager.kt
package com.cere.csplayer.content

import android.content.Context
import com.cere.csplayer.Constants
import com.cere.csplayer.OnPermissionCallback
import com.cere.csplayer.data.AppDatabase
import com.cere.csplayer.data.SharePre
import com.cere.csplayer.entity.Music
import com.cere.csplayer.entity.Play
import com.cere.csplayer.until.PermissionUtils
import com.cere.csplayer.until.Utils
import com.hjq.permissions.Permission
import com.hjq.permissions.XXPermissions
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.schedulers.Schedulers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

/**
 * Created by CheRevir on 2020/9/9
 */
class MusicManager(private val context: Context) :
    MusicScanner.OnMusicScanDoneListener {
    lateinit var musics: List<Music>
    lateinit var plays: List<Play>

    override fun onMusicScanDone(list: List<Music>) {
        if (list.isNotEmpty()) {
            Observable.create<List<Music>> { ob ->
                ob.onNext(Utils.deepCopy(musics))
            }.subscribeOn(Schedulers.newThread())
                .subscribe {
                    if (it.isEmpty()) {
                        GlobalScope.launch {
                            AppDatabase.instance.getMusicDao().insert(list)
                            AppDatabase.instance.getPlayDao().deleteAll()
                            AppDatabase.instance.getPlayDao().insert(
                                musicsToPlays(
                                    list,
                                    Play(list[0].id)
                                )
                            )
                        }
                    } else {
                        GlobalScope.launch {
                            val removeMusics = listCompare(list, it)
                            val removePlays = ArrayList<Play>(removeMusics.size)
                            removeMusics.forEach { music ->
                                removePlays.add(plays[plays.indexOf(Play(music.id))])
                            }
                            if (removeMusics.isNotEmpty()) AppDatabase.instance.getMusicDao()
                                .delete(removeMusics)
                            if (removePlays.isNotEmpty()) AppDatabase.instance.getPlayDao()
                                .delete(removePlays)
                        }
                        GlobalScope.launch {
                            val addMusics = listCompare(it, list)
                            val addPlays = ArrayList<Play>(addMusics.size)
                            addMusics.forEach { music ->
                                addPlays.add(Play(music.id))
                            }
                            if (addMusics.isNotEmpty()) AppDatabase.instance.getMusicDao()
                                .insert(addMusics)
                            if (addPlays.isNotEmpty()) AppDatabase.instance.getPlayDao()
                                .insert(addPlays)
                        }
                    }
                }
        }
    }

    suspend fun buildPlays(position: Int) {
        val list = musicsToPlays(musics, Play(position))
        AppDatabase.instance.getPlayDao().deleteAll()
        AppDatabase.instance.getPlayDao().insert(list)
    }

    private fun musicsToPlays(list: List<Music>, play: Play): List<Play> {
        val plays = ArrayList<Play>(list.size)
        list.forEach {
            plays.add(Play(it.id))
        }
        if (SharePre.getBoolean(Constants.MODE_SHUFFLE, false)) {
            plays.shuffle()
            plays[plays.indexOf(play)] = plays[0]
            plays[0] = play
        }
        return plays
    }

    private fun listCompare(list1: List<Music>, list2: List<Music>): List<Music> {
        val list = ArrayList<Music>()
        val map = HashMap<Int, Int>(list1.size + list2.size)
        list1.forEach {
            map[it.id] = 1
        }
        list2.forEach {
            if (map[it.id] != 1) {
                list.add(it)
            }
        }
        return list
    }

    @Suppress("DEPRECATION")
    fun scan() {
        if (XXPermissions.hasPermission(context, Permission.READ_EXTERNAL_STORAGE)) {
            MusicScanner(context, this).execute()
        } else {
            PermissionUtils.instance.getXXPermissions()
                ?.permission(Permission.READ_EXTERNAL_STORAGE)
                ?.request(
                    object : OnPermissionCallback(context) {
                        override fun hasPermission(granted: MutableList<String>, all: Boolean) {
                            scan()
                        }
                    })
        }
    }
}